QuickEditor: AltText - Save the avatars list locally in the AvatarsRespository instead of in the `AvatarPickerViewModel`

hamorillo opened the pull request:
Closes #516 

### Description

With the new `AltTextPage`, we need to save the avatar's data in a shared place as we need to access that info from both screens. We already have the `AvatarsRepository`, so that's the best place to keep the avatars list locally. We need to create a flow for each email and update that flow in every repository operation.

We were doing a similar thing in the `AvatarPickerViewModel`. We had a variable where we kept the avatars list and updated it in the different operations (avatar upload, deletion, selection...). This PR moves that logic to the repository while sharing the avatar flow for each email. 

### Testing Steps

Everything should work as before except the AltText, where this PR should solve the issue of reopening the AltText page after changing it. In that situation, the new AltText should be visible, not as it was before when the old AltText was visible.

- Smoke test all the operations in the QE (avatar upload, deletion, selection, rating)
- Test that after changing the `AltText` and reopening the AltTextPage for the same image, the new AltText is visible instead of the old one.

https://github.com/user-attachments/assets/9257df1d-8c47-410a-8717-df6c2cf61b52

hamorillo committed:
First step to keep a single source of truth for the avatars in the AvatarsRepository

AvatarRepository should be a singleton to be shared between different places.

hamorillo committed:
Update avatarsFlow after avatar selection

hamorillo committed:
Update avatarsFlow after avatar upload

hamorillo committed:
Update avatarsFlow after avatar deletion

hamorillo committed:
Extract methods to update avatars locally

hamorillo commented on gravatar-quickeditor/src/main/java/com/gravatar/quickeditor/QuickEditorContainer.kt:
```diff
@@ -91,12 +91,13 @@ internal class QuickEditorContainer private constructor(
 
-    val avatarRepository: AvatarRepository
-        get() = AvatarRepository(
+    val avatarRepository: AvatarRepository by lazy {
```
The `avatarRepository` needs to be a singleton that can be shared between screens.

AdamGrzybkowski replied:
I think we should create `AvatarsStorage` to offload the repository. That way only the storage needs to be a Singleton and the Repository will make sure we get the latest data (from remote or local).

hamorillo replied:
Good point @AdamGrzybkowski. I agree that creating `AvatarsStorage` will make the `AvatarRepository` cleaner. I'll refactor this.

hamorillo replied:
I've refactored the code as discussed. `AvatarsStorage` handles the avatar list for each email locally.

AdamGrzybkowski commented on gravatar-quickeditor/src/main/java/com/gravatar/quickeditor/ui/avatarpicker/AvatarPickerViewModel.kt:
```diff
@@ -443,6 +428,16 @@ internal class AvatarPickerViewModel(
             .launchIn(viewModelScope)
     }
 
+    private fun avatarsObserver() {
```
⛏️ I would rename it to `collectAvatars` or `observe` - the current name sounds like we would return or create the Observer.

AdamGrzybkowski commented on gravatar-quickeditor/src/main/java/com/gravatar/quickeditor/data/repository/AvatarRepository.kt:
```diff
@@ -4,30 +4,42 @@ import android.net.Uri
-                    )
+                        )
+                    }
+                    avatarsFlow(email).emit(emailAvatars)
+                    GravatarResult.Success(emailAvatars)
```
💡  If we are going to have a separate flow with avatars I think we should rename this function to `refreshAvatars` and not return them here. This could be confusing, and unclear whether the avatars returned should be used or those emitted in the flow.

hamorillo replied:
I renamed the function; however, it still returns the avatar lists because we use it to change the scroll position in the viewmodel. If we don't return the avatar list here we need to refactor how we scroll to the selected avatar in that case. I'm not saying I don't want to do it, but I would prefer to work on it in a different PR because this one is growing quite a bit and doesn't look like a blocker to me. Wdyt?

AdamGrzybkowski replied:
hmm, I see. I think we can do it later. We will have two places where we get the data, so I see this as a potential source for problems in the future.

hamorillo replied:
> We will have two places where we get the data, so I see this as a potential source for problems in the future.

Yeah, I agree.

AdamGrzybkowski commented on gravatar-quickeditor/src/main/java/com/gravatar/quickeditor/data/repository/AvatarRepository.kt:
```diff
@@ -4,30 +4,42 @@ import android.net.Uri
     private val tokenStorage: TokenStorage,
     private val dispatcher: CoroutineDispatcher,
 ) {
+    fun getAvatarsFlow(email: Email) = avatarsFlow(email).asSharedFlow()
```
```suggestion
    fun getAvatars(email: Email) = avatarsFlow(email).asSharedFlow()
```

⛏️ I think I would shorten the name and skip the `flow` suffix. The return type is specified so not sure having it in the name adds much value.

AdamGrzybkowski commented on gravatar-quickeditor/src/main/java/com/gravatar/quickeditor/data/repository/AvatarRepository.kt:
```diff
@@ -4,30 +4,42 @@ import android.net.Uri
 ) {
+    fun getAvatarsFlow(email: Email) = avatarsFlow(email).asSharedFlow()
+
+    private val avatarsFlows = mutableMapOf<String, MutableSharedFlow<EmailAvatars>>()
```
The `EmailAvatars` object was created based on how the initial API was structured but also to serve the presentation layer. I don't think this is how we should store it. The `Avatar` object already has the `selected` field so it should be enough to store a plain list of Avatars and mapped to the `EmailAvatars` in the ViewModel.

AdamGrzybkowski reviewed (commented):
Works great! But I would like to hear from you what you think about the comment about extracting all the local storage from the repository. The Repository would stay cleaner and without any knowledge of how objects are stored. It would make sure the data is returned from a proper source.

hamorillo committed:
Extract to AvatarStorage the logic to keep the avatars locally for a given email

hamorillo committed:
Remove "EmailAvatars" from AvatarRepository and AvatarStorage

EmailAvatars was created to be used in the ViewModel/View and shouldn't be used in the Repository or LocalStorage.

hamorillo committed:
Rename avatarsObserver to collectAvatars

hamorillo committed:
Rename AvatarRepository.getAvatars to refreshAvatars

hamorillo committed:
Rename getAvatarsFlow to getAvatars

hamorillo commented on gravatar-quickeditor/src/main/java/com/gravatar/quickeditor/data/repository/AvatarRepository.kt:
```diff
@@ -84,14 +95,13 @@ internal class AvatarRepository(
     }
 }
-
-internal data class EmailAvatars(
```
`AvatarRepository` will work only with `Avatar`s, the viewmodel will map the Avatars to EmailAvatars, so moving this class to the PickersViewModel.

AdamGrzybkowski commented on gravatar-quickeditor/src/main/java/com/gravatar/quickeditor/data/storage/AvatarStorage.kt:
```diff
@@ -0,0 +1,86 @@
+    internal fun avatarsFlow(email: Email): MutableSharedFlow<List<Avatar>> =
+        avatarsFlows.getOrPut(email.hash().toString()) { MutableSharedFlow(replay = 1) }
+
+    internal suspend fun storeAvatars(avatars: List<Avatar>, email: Email) {
```
⛏️ Totall nitpick, but email param is first in other functions :D

AdamGrzybkowski commented on gravatar-quickeditor/src/main/java/com/gravatar/quickeditor/ui/avatarpicker/AvatarPickerViewModel.kt:
```diff
@@ -51,6 +51,7 @@ internal class AvatarPickerViewModel(
     init {
         refresh()
         nonAvatarSelectedAlertObserver()
+        collectAvatars()
```
💡 Not sure but I think we might be able to combine these two collectors now. It's not something for this PR though.

hamorillo replied:
Issue #522

AdamGrzybkowski commented on gravatar-quickeditor/src/main/java/com/gravatar/quickeditor/data/storage/AvatarStorage.kt:
```diff
@@ -0,0 +1,86 @@
+        avatarsFlow(email).emit(avatars)
+    }
+
+    internal suspend fun markAvatarAsSelected(email: Email, avatarId: String) {
```
```suggestion
    internal suspend fun selectAvatar(email: Email, avatarId: String) {
```

AdamGrzybkowski commented on gravatar-quickeditor/src/test/java/com/gravatar/quickeditor/data/storage/AvatarStorageTest.kt:
```diff
@@ -0,0 +1,128 @@
+    }
+
+    @Test
+    fun `given avatars when storeAvatars then avatarsFlow emits a avatars`() = runTest {
```
```suggestion
    fun `given avatars when storeAvatars then avatarsFlow emits avatars`() = runTest {
```

AdamGrzybkowski commented on gravatar-quickeditor/src/test/java/com/gravatar/quickeditor/data/storage/AvatarStorageTest.kt:
```diff
@@ -0,0 +1,128 @@
+    }
+
+    @Test
+    fun `given avatarId when markAvatarAsSelected then avatarsFlow emits a avatars with selected avatar`() = runTest {
```
```suggestion
    fun `given avatarId when markAvatarAsSelected then avatarsFlow emits avatars with selected avatar`() = runTest {
```

AdamGrzybkowski commented on gravatar-quickeditor/src/test/java/com/gravatar/quickeditor/data/storage/AvatarStorageTest.kt:
```diff
@@ -0,0 +1,128 @@
+    }
+
+    @Test
+    fun `given avatar when addAvatar then avatarsFlow emits a avatars with updated avatar`() = runTest {
```
```suggestion
    fun `given avatar when addAvatar then avatarsFlow emits avatars`() = runTest {
```

This one adds new one, not updates it, right?

AdamGrzybkowski commented on gravatar-quickeditor/src/test/java/com/gravatar/quickeditor/data/storage/AvatarStorageTest.kt:
```diff
@@ -0,0 +1,128 @@
+    }
+
+    @Test
+    fun `given avatarId when deleteAvatar then avatarsFlow emits a avatars without deleted avatar`() = runTest {
```
```suggestion
    fun `given avatarId when deleteAvatar then avatarsFlow emits avatars without deleted avatar`() = runTest {
```

AdamGrzybkowski commented on gravatar-quickeditor/src/test/java/com/gravatar/quickeditor/data/storage/AvatarStorageTest.kt:
```diff
@@ -0,0 +1,128 @@
+    }
+
+    @Test
+    fun `given avatar when updateAvatar then avatarsFlow emits a avatars with updated avatar`() = runTest {
```
```suggestion
    fun `given avatar when updateAvatar then avatarsFlow emits avatars with updated avatar`() = runTest {
```

AdamGrzybkowski commented on gravatar-quickeditor/src/test/java/com/gravatar/quickeditor/ui/avatarpicker/AvatarPickerViewModelTest.kt:
```diff
@@ -351,69 +351,8 @@ class AvatarPickerViewModelTest {
-
-    @OptIn(ExperimentalCoroutinesApi::class)
-    @Test
-    fun `given avatar returned when avatar with the same id then uiState is updated`() = runTest {
```
❓ This one is not needed now?

hamorillo replied:
This logic is no longer part of the ViewModel as it's handled by the AvatarStore/Repository, so I think this test is unnecessary.

AdamGrzybkowski reviewed (approved):
It works great, and the alt text is synchronized! 

We could slim down the amount of data we pass to the AltTextPage now.

hamorillo committed:
Switch params order in AvatarStorage.selectAvatar

hamorillo committed:
Fix typos

Co-authored-by: AdamGrzybkowski <agrzybkowski@outlook.com>

hamorillo commented:
> We could slim down the amount of data we pass to the AltTextPage now.

Issue #523
